stellargraph/layer/appnp.py

```diff
@@ -0,0 +1,244 @@
+        self._layers.append(
+            Lambda(lambda inputs:
+               K.expand_dims(
+                   K.gather(
+                       K.squeeze(inputs[0], 0),
+                       tf.cast(K.squeeze(inputs[1], 0), tf.int32)
+                    ),
+                    0
+               )
+            )
+        )
+
+    def __call__(self, x):
+        """
+        Apply APPNP to the inputs.
+        The input tensors are expected to be a list of the following:
+        [
+            Node features shape (1, N, F),
+            Adjacency indices (1, E, 2),
+            Adjacency values (1, E),
+            Output indices (1, O)
+        ]
+        where N is the number of nodes, F the number of input features,
+              E is the number of edges, O the number of output nodes.
+        Args:
+            x (Tensor): input tensors
+        Returns:
+            Output tensor
+        """
+        x_in, out_indices, *As = x
+
+        # Currently we require the batch dimension to be one for full-batch methods
+        batch_dim, n_nodes, _ = K.int_shape(x_in)
+        if batch_dim != 1:
+            raise ValueError(
+                "Currently full-batch methods only support a batch dimension of one"
+            )
+
+        # Convert input indices & values to a sparse matrix
+        if self.use_sparse:
+            A_indices, A_values = As
+            Ainput = [
+                SqueezedSparseConversion(
+                    shape=(n_nodes, n_nodes), dtype=A_values.dtype
+                )([A_indices, A_values])
+            ]
+
+        # Otherwise, create dense matrix from input tensor
+        else:
+            Ainput = [Lambda(lambda A: K.squeeze(A, 0))(A) for A in As]
+
+        # TODO: Support multiple matrices?
+        if len(Ainput) != 1:
+            raise NotImplementedError(
+                "The APPNP method currently only accepts a single matrix"
+            )
+
+        h_layer = x_in
+        for layer in self._layers[:-1]:
+            if isinstance(layer, GraphConvolution):
+                # For a GCN layer add the matrix and output indices
+                # Note that the output indices are never used in these layers - `final_layer=False`
+                h_layer = layer([h_layer, out_indices] + Ainput)
+                # propogate with the feature layer
+                h_layer = Lambda(lambda x:
+                                    (1 - self.transport_probability) * x[0] + self.transport_probability * x[1]
+                                 )([h_layer, feature_layer])
+            elif isinstance(layer, Dense):
+                h_layer = layer(h_layer)
+                # store feature layer for propogation
+                feature_layer = h_layer
+            else:
+                # For other (non-graph) layers only supply the input tensor
+                h_layer = layer(h_layer)
+
+        #apply gather layer
+        h_layer = self._layers[-1]([h_layer, out_indices])
```

Comment:
You could also just create and apply the gather layer here, not all layers used in the model need to be stored in `._layers`.

Reply:
Good idea, I'll move the gather layer over.